## CS-E4825 - Probabilistic Machine Learning D (spring 2025)

Pekka Marttinen, Negar Safinianaini, Mihriban Kocak, Bo Zheng, Batuhan Avci.

## Exercise 3, due on Friday 31st January at 10:15.

### Contents
1. Problem 1: Poisson-Gamma
2. Problem 2: Multivariate Gaussian
3. Problem 3: Posterior of regression weights

## Problem 1: Poisson-Gamma

Suppose you have $N$ i.i.d. observations $\mathbf{x}= \{x_i\}_{i=1}^N$ from a $\operatorname{Poisson}(\lambda)$ distribution with a rate parameter $\lambda$ that has a conjugate prior 

$$\lambda \sim \operatorname{Gamma}(a,b)$$

with the shape and rate hyperparameters $a$ and $b$. Derive the posterior distribution $\lambda|\bf{x}$.

Write your solutions in LateX or attach a picture in the answer cell provided below. You can add a picture using the command ```!(imagename_in_the_folder.jpg)```. Latex in here works similarly as you would write it normally! You can use some of the definitions from the exercise description as a reference. The list of valid Latex commands in Jypyter notebook can be found here: http://www.onemathematicalcat.org/MathJaxDocumentation/TeXSyntax.htm




Let's start out by defining the necessary probability distributions:

Prior: $P(\lambda) = \frac{b^a}{\Gamma(a)} \lambda^{a-1} e^{-b \lambda}$ ($\lambda \sim \text{Gamma}(a,b)$)

Marginal probability: $P(x) = \int_0^\infty P(x \ | \ \lambda = u) \ du = \int_0^\infty \Pi_{i=1}^N \frac{u^{x_i} e^{-u}}{(x_i)!} du $ due to the fact that the observations in $x$ are i.i.d.

Likelihood of $x$ given $\lambda$: $P(x \ | \ \lambda) = \Pi_{i=1}^N \frac{\lambda^{x_i} e^{-\lambda}}{(x_i)!}$

Let us look at the numerator more closely (likelihood times prior):

$P(\lambda \ | \ x) \propto P(x \ | \ \lambda) P(\lambda) = \Pi_{i=1}^N ( \frac{\lambda^{x_i} e^{-\lambda}}{(x_i)!}) \cdot \frac{b^a}{\Gamma(a)} \lambda^{a-1} e^{-b \lambda} $

$P(\lambda \ | \ x) \propto ( \frac{\lambda^{\sum_{i=1}^N x_i} e^{-\lambda N}}{\Pi_{i=1}^N (x_i)!}) \cdot \frac{b^a}{\Gamma(a)} \lambda^{a-1} e^{-b \lambda} $

$P(\lambda \ | \ x) \propto \lambda^{a - 1 + \sum_{i=1}^N x_i} \ e^{-(b + N) \lambda} $ with the constants removed

$\Rightarrow \lambda \ | \ x \sim \text{Gamma}(a + \sum_{i=1}^N x_i, b + N) $ 

$\lambda \ | \ x$ is Gamma distributed with parameters $a + \sum_{i=1}^N x_i$ and $b+N$


## Problem 2: Multivariate Gaussian

Suppose we have $N$ i.i.d. observations $\mathbf{X} = \{\mathbf{x}_i\}_{i=1}^N$ from a multivariate Gaussian distribution $$\mathbf{x}_i \mid \boldsymbol{\mu} \sim \mathcal{N}(\boldsymbol{\mu}, \boldsymbol{\Sigma})$$ with unknown mean parameter $\boldsymbol{\mu}$  and a known covariance matrix $\boldsymbol{\Sigma}$. As prior information on the mean parameter we have $$ \boldsymbol{\mu} \sim \mathcal{N}(\mathbf{m_0}, \mathbf{S_0}). $$

__(a)__ Derive the posterior distribution $p(\boldsymbol{\mu}|\mathbf{X})$ of the mean parameter $\boldsymbol{\mu}$. Write your solution in LateX or attach a picture of the solution in the cell below.

__(b)__ Compare the Bayesian estimate (posterior mean) to the maximum likelihood estimate by generating $N=10$ observations from the bivariate Gaussian 
        $$\mathcal{N}\left(\begin{bmatrix}0 \\ 0\end{bmatrix}, \begin{bmatrix}1 & 0 \\ 0 & 1\end{bmatrix}\right).$$
For this you can use the Python function [numpy.random.normal](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.normal.html), making use of the fact that the elements of the bivariate random vectors are independent. In the Bayesian case, use the prior with $\mathbf{m_0} = [0,0]^T$ and $\mathbf{S_0} = [\begin{smallmatrix}0.1 & 0 \\ 0 & 0.1\end{smallmatrix}]$. Report both estimates. Is the Bayesian estimate closer to the true value $\boldsymbol{\mu} = [0,0]^T$? Use the code template given below (after the answer cell) to complete your answer.

Write your solutions to __(a)__ and __(b)__ in LateX or attach a picture in the answer cell provided below. 



Part a)

Assuming that the covariance matrix $S_0 \in \mathbb{R}^{k \times k}$ is a positive definite matrix (it is invertible among other nice things), and when $\mu \in \mathbb{R}^k$ is the mean vector. We also know that the resulting posterior distribution should be gaussian given that the likelihood and prior are both gaussian.

Prior: $p(\mu) = \frac{1}{\sqrt{(2 \pi)^k \text{det}(S_0)}} e^{-\frac{1}{2} (\mu - m_0)^\top S_0^{-1}(\mu - m_0)}$

Likelihood of $X$ given $\mu$: $p(X \ | \ \mu) = \Pi_{i=1}^N \frac{1}{\sqrt{(2 \pi)^k \text{det}(\Sigma)}} e^{-\frac{1}{2} (x_i - \mu)^\top \Sigma^{-1}(x_i - \mu)} = (\frac{1}{\sqrt{(2 \pi)^k \text{det}(\Sigma)}})^N \exp(\sum_{i=1}^N(-\frac{1}{2} (x_i - \mu)^\top \Sigma^{-1}(x_i - \mu)))$

$p(X \ | \ \mu) = (\frac{1}{\sqrt{(2 \pi)^k \text{det}(\Sigma)}})^N \exp(-\frac{1}{2} \sum_{i=1}^N x_i^\top \Sigma^{-1}x_i  - 2 N \mu^\top \Sigma^{-1} \bar{x} + N \mu^\top \Sigma^{-1} \mu)$

Let us now look at the likelihood times the prior:

$p(X \ | \ \mu) \ p(\mu) \propto \exp(- \frac{1}{2} (- 2 N \mu^\top \Sigma^{-1} \bar{x} + N \mu^\top \Sigma^{-1} \mu) \  \text{exp}(-\frac{1}{2} (\mu - m_0)^\top S_0^{-1}(\mu - m_0))$

$p(X \ | \ \mu) \ p(\mu) \propto \text{exp}(-\frac{1}{2} (- 2 N \mu^\top \Sigma^{-1} \bar{x} + N \mu^\top \Sigma^{-1} \mu + \mu^\top S_0^{-1} \mu - 2 m_0^\top S_0^{-1} \mu + m_0^\top S_0^{-1} m_0) )$

$p(X \ | \ \mu) \ p(\mu) \propto \text{exp}(-\frac{1}{2} (\mu^\top (S_0^{-1} + N \Sigma^{-1}) \mu - 2 \mu^\top (- N \Sigma^{-1} \bar{x} - S_0^{-1} m_0))$

$p(X \ | \ \mu) \ p(\mu) \propto  \text{exp}(-\frac{1}{2} (\mu - (S_0^{-1} + N \Sigma^{-1})^{-1}( S_0^{-1} m_0 + N \Sigma^{-1} \bar{x}) ))^\top (S_0 + N \Sigma^{-1}) (\mu - (S_0^{-1} + N \Sigma^{-1})^{-1}( S_0^{-1} m_0 + N \Sigma^{-1} \bar{x}) )   )$

Now we notice that the posterior is indeed proportional to a multivariate normal distribution with parameters:


$\Sigma_{\text{new}} = (S_0^{-1} + N \Sigma^{-1})^{-1} $ and $\mu_{\text{new}} = \Sigma_{\text{new}} ( S_0^{-1} m_0 + N \Sigma^{-1} \bar{x}) $

Thus: $p(X \ | \ \mu) \ p(\mu) \propto  \text{exp}(-\frac{1}{2} (\mu - \mu_{new})^\top \Sigma_{new}^{-1}) (\mu - \mu_{new} )) \Rightarrow X \ | \ \mu \sim \mathcal{N}(\mu_{new}, \Sigma_{new})$

Note in the derivation above I did not bother to list out all of the constants, since they are all absorbed in the normalizing constant.

In [7]:
# template for 2(b)
import numpy as np
from numpy.linalg import inv

m_0 = np.array([0,0])
S0 = np.array([[0.1, 0],[0, 0.1]])
Sigma = np.array([[1, 0],[0, 1]])
N = 10

# Sample N bivariate normal vectors
# compute MLE and also the posterior mean solution

# x = ? #EXERCISE
# mle = ? #EXERCISE
# posterior_mean = ? #EXERCISE 

# YOUR CODE HERE

# MLE
mu = np.array([0,0])

x_sample = np.column_stack((np.random.normal(mu[0], Sigma[0,0], N), np.random.normal(mu[1], Sigma[1,1], N)))

mle = np.mean(x_sample, axis=0)


# Bayesian
Sigma_inv = inv(Sigma)
S0_inv = inv(S0)

Sigma_new = inv(S0_inv + N * Sigma_inv)
mu_new = Sigma_new @ (S0_inv @ m_0 + N * S0_inv @ mle)

posterior_sample = np.column_stack((np.random.normal(mu_new[0], Sigma_new[0,0], N), np.random.normal(mu_new[1], Sigma_new[1,1], N)))

posterior_mean = np.mean(posterior_sample, axis=0)

print("Posterior mean: ", posterior_mean)
print("MLE: ", mle)


Posterior mean:  [-1.43304735  1.58104615]
MLE:  [-0.28360705  0.31824903]


In the above code the required samplings and computations are done and based on the results the MLE gets closer to the true value of $\mu = \begin{bmatrix} 0 & 0 \end{bmatrix}^\top$. The posterior mean was: $\begin{bmatrix} -1.4330 & 1.5810 \end{bmatrix}^\top$ and the MLE was $\begin{bmatrix} -0.2836 & 0.3182 \end{bmatrix}^\top$

In [55]:
print((10 * 0.3953)/20)
print(10 * 0.5877/20)

0.19765
0.29385


# Problem 3: Posterior of regression weights

Suppose $y_{i}=\mathbf{w}^{T}\mathbf{x}_{i}+\epsilon_{i},$ for $i=1,\ldots,n,$ where $\epsilon_{i}\sim \mathcal{N}(0,\beta^{-1})$. Assume a prior $$\mathbf{w} \sim \mathcal{N} (\mathbf{0},\alpha^{-1}\mathbf{I}).$$ Use 'completing the square' to show that the posterior of $\mathbf{w}$ is given by $p(\mathbf{w} \mid \mathbf{y}, \mathbf{x}, \alpha, \beta)=\mathcal{N}(\mathbf{w} \mid \mathbf{m}, \mathbf{S}),$ where 
\begin{align*}
    \mathbf{S} &= \left( \alpha \mathbf{I} + \beta \sum_{i=1}^n \mathbf{x}_i \mathbf{x}_i^T \right)^{-1}\;, \\
    \mathbf{m} &= \beta \mathbf{S} \sum_{i=1}^{n} y_i \mathbf{x}_i.
\end{align*}

Write your solution in LateX or attach a picture of the solution in the cell below.


The prior: 

$p(w) \propto \text{exp}(-\frac{1}{2} w^\top (\alpha^{-1}I)^{-1} w) = \text{exp}(-\frac{1}{2} w^\top (\alpha I) w)$

We know that $y_i$ given $w, x, \alpha, \beta$ follows a normal distribution s.t. 

$y_i \ | \ w, x, \alpha, \beta \sim \mathcal{N}(w^\top x_i, \beta^{-1})$

Therefore the likelihood is proportional to: 

$p(y_i \ | \ w, x, \alpha, \beta) \propto \text{exp}(-\frac{1}{2\beta} (y_i - w^\top x_i)^2) $

And using that and the assumption of independence between $y_i$'s: 

$p(y \ | \ w, x, \alpha, \beta) \propto \Pi_{i=1}^N \text{exp}(-\frac{\beta}{2} (y_i - w^\top x_i)^2) $

$p(y \ | \ w, x, \alpha, \beta) \propto \text{exp}(-\frac{\beta}{2} \sum_{i=1}^N (y_i - w^\top x_i)^2) $

Now the likelihood times the prior: 

$p(y \ | \ w, x, \alpha, \beta) \ p(w) \propto \text{exp}(-\frac{\beta}{2} \sum_{i=1}^N (y_i - w^\top x_i)^2) \ \text{exp}(-\frac{1}{2} w^\top (\alpha I) w)$

$p(y \ | \ w, x, \alpha, \beta) \ p(w) \propto \text{exp}(-\frac{1}{2} w^\top (\alpha I) w - \frac{\beta}{2} \sum_{i=1}^N (y_i - w^\top x_i)^2)$

Let's look at just the exponent:

$-\frac{1}{2} w^\top (\alpha I) w - \frac{\beta}{2} \sum_{i=1}^N (y_i - w^\top x_i)^2 $

$= -\frac{1}{2} w^\top (\alpha I) w - \frac{\beta}{2} \sum_{i=1}^N y_i^2  - 2 w^\top x_i y_i + (w^\top x_i)(w^\top x_i) $

$= -\frac{1}{2} w^\top (\alpha I) w - \frac{\beta}{2} \sum_{i=1}^N y_i^2  - 2 w^\top x_i y_i + (w^\top (x_i x_i^\top) w)$

$= -\frac{1}{2} ( w^\top (\alpha I) w + \beta (\sum_{i=1}^N y_i^2  - 2 w^\top \sum_{i=1}^N x_i y_i + (w^\top (\sum_{i=1}^N x_i x_i^\top) w) )  )$

$= -\frac{1}{2} ( w^\top (\alpha I + \beta \sum_{i=1}^N x_i x_i^\top) w + \beta (\sum_{i=1}^N y_i^2  - 2 w^\top \sum_{i=1}^N y_i x_i )  )$

$= -(\frac{1}{2} w^\top (\alpha I + \beta \sum_{i=1}^N x_i x_i^\top) w - \beta (\sum_{i=1}^N y_i x_i )^\top w)) - \frac{\beta}{2} (\sum_{i=1}^N y_i^2 )$

Now the first two terms are in the form of: $\frac{1}{2} x^\top A x - b^\top x$ for which we can use completing the square to get:  $\frac{1}{2} (x - A^{-1} b)^\top A (x - A^{-1} b) - \frac{1}{2} b^\top A^{-1} b $ for symmetric matrices $A = A^\top$.

With this we get that the exponent is equal to (denote $S = (\alpha I + \beta \sum_{i=1}^N x_i x_i^\top)^{-1}$ and $m = \beta S \sum_{i=1}^N y_i x_i$)

$= -\frac{1}{2} (w-m)^\top S^{-1} (w-m) + \frac{\beta}{2} m^\top S m - \frac{\beta}{2} (\sum_{i=1}^N y_i^2 )$

Now notice that the last two terms are constant w.r.t. to $w$. Thus we can say that:

$p(y \ | \ w, x, \alpha, \beta) \ p(w) \propto \text{exp}(-\frac{1}{2} (w-m)^\top S^{-1} (w-m))$

Therefore since the posterior is proportional to likelihood times prior, indeed $w \ | \ y, x, \alpha, \beta \sim \mathcal{N}(m, S)$, where $S = (\alpha I + \beta \sum_{i=1}^N x_i x_i^\top)^{-1}$ and $m = \beta S \sum_{i=1}^N y_i x_i$
